skxray/correlation.py

```diff
@@ -0,0 +1,299 @@
+    G = np.zeros(((num_levels + 1)*num_bufs/2, num_qs),
+                 dtype=np.float64)
+    # matrix of past intensity normalizations
+    IAP = np.zeros(((num_levels + 1)*num_bufs/2, num_qs),
+                   dtype=np.float64)
+    # matrix of future intensity normalizations
+    IAF = np.zeros(((num_levels + 1)*num_bufs/2, num_qs),
+                   dtype=np.float64)
+
+    # matrix of one-time correlation for required roi's
+    g2 = np.zeros((num_levels, num_qs), dtype=np.float64)
+
+    # correlation for delays, images must be keep for up to maximum
+    # delay in buf
+    buf = np.zeros((num_levels, num_bufs, np.sum(num_pixels)),
+                   dtype=np.float64)
+
+    cts = np.zeros(num_levels)
+    cur = np.ones((num_levels), dtype=np.int64)
+
+    num = np.array(np.zeros(num_levels), dtype=np.int64)
+
+    # number of frames(images)
+    num_frames = img_stack.shape[0]
+
+    start_time = time.time()
+    for n in range(0, num_frames):  # changed the number of frames
+        image_array = img_stack[n]
+
+        cur[0] = 1 + cur[0] % num_bufs  # increment buffer
+
+        buf[0, cur[0] - 1] = (np.ravel(image_array))[pixel_list]
+        G, IAP, IAF, num = _process(buf, G, IAP, IAF, q_inds,
+                                    num_bufs, num_pixels, num, level=0,
+                                    buf_no=cur[0] - 1)
+        if num_levels > 1:
+            processing = 1
+            level = 1
+        else:
+            processing = 0
+
+        while processing:
+            if cts[level]:
+                prev = 1 + (cur[level - 1] - 2 + num_bufs) % num_bufs
+                cur[level] = 1 + cur[level] % num_bufs
+                buf[level, cur[level] - 1] = (buf[level - 1, prev - 1] +
+                                              buf[level - 1,
+                                                  cur[level - 1] - 1])/2
+
+                cts[level] = 0
+                G, IAP, IAF, num = _process(buf, G, IAP, IAF, q_inds,
+                                            num_bufs, num_pixels, num,
+                                            level=level, buf_no=cur[level]-1,)
+                level += 1
+
+                # Checking whether there is next level for processing
+                if level < num_levels:
+                    processing = 1
+                else:
+                    processing = 0
+            else:
+                cts[level] = 1
+                processing = 0
+
+    elapsed_time = time.time() - start_time
+
+    if len(np.where(IAP == 0)[0]) != 0:
+        g_max = np.where(IAP == 0)[0][0]
+    else:
+        g_max = IAP.shape[0]
+
+    g2 = (G[: g_max] / (IAP[: g_max] * IAF[: g_max]))
+
+    tot_channels, lag_steps = core.multi_tau_lags(num_levels,
+                                                  num_bufs)
+
+    lag_steps = lag_steps[:g2.shape[0]]
+
+    return g2, lag_steps, elapsed_time
+
+
+def _process(buf, G, IAP, IAF, q_inds, num_bufs,
```

Comment:
What does this function get you over embedding this code directly in the loop above?